### PR TITLE
[Ray Core] Adds in Google Cloud TPUs as a native Resource

### PR DESCRIPTION
**Pull request by allenwang28**
## Why are these changes needed?

The issue below has more details, but at a high level this change addresses the feature request of adding in TPUs as a native resource within Ray.

To a user, these changes intend to shift from:
```
@ray.remote(resources={"TPU": 1})
def my_function():
    ...
```
to 
```
@ray.remote(num_tpus=4)
def my_function():
    ...
```
Since we're adding TPUs as a native resource, this gives users the added ability to access individual TPU chips in a TPU host, e.g. the following would be valid:
```
@ray.remote(num_tpus=1) # or 2
def my_function():
    import jax
    print(jax.device_count()) -> 1 (or 2)
```

This is enabled by setting environment variables as described in https://github.com/google/jax/issues/14977.

Short overview of changes:
- Add in autodetect num TPU chips + TPU version (+ tests for each)
- Adds in `num_tpus` in `ray_params` and does input sanitization similar to how GPUs function (e.g. `num_tpus` instead of `TPU` in resources)
- Adds in several constants
- Since there are a lot of changes related to neuron cores (from https://github.com/ray-project/ray/pull/37998) I tried to consolidate the neuron core specific logic with TPU logic where applicable, for instance changing `_validate_neuron_core_accelerator` -> `_validate_accelerator` in `ray_option_utils.py` 
- And added in a lot of tests!

## Related issue number

https://github.com/ray-project/ray/issues/38085

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by architkulkarni** [comment]
@richardliaw @scv119 Do you approve the API change? I'll review the PR, perhaps you can suggest other reviewers too.

**Comment by ericl** [comment]
I don't believe we are intending to add vendor-specific resources directly to the Ray API--- the pattern followed should be similar to https://github.com/ray-project/ray/pull/37998 for all new accelerator types.

**Comment by allenwang28** [comment]
> I don't believe we are intending to add vendor-specific resources directly to the Ray API--- the pattern followed should be similar to #37998 for all new accelerator types.

Thanks - my mistake. I'll pull out the API specific changes and align with the `neuron_cores` approach.

**Comment by allenwang28** [comment]
I've updated the PR to remove `num_tpus`, PTAL!

File: src/ray/common/ray_config_def.h
Comment (architkulkarni) [inline src/ray/common/ray_config_def.h:691]:
```suggestion
/// Default custom_unit_instance_resources is "neuron_cores,TPU".
```

Is this right? Feel free to rewrite the whole comment if you understand it, it's kind of confusing to me at the moment.

**Review by architkulkarni — APPROVED** [review]
Looks good to me! Some nits around maintenance/documentation in case someone less familiar with TPU has to update this code in the future, but these don't need to block the PR in my opinion

**Comment by architkulkarni** [comment]
Assigning @scv119 as ray core codeowner, and someone who reviewed the AWS neuron core PR

**Comment by jjyao** [comment]
Hi @allenwang28,

Thanks for the contribution. Given multiple teams are trying to add support of different accelerators to Ray. We'd like to come up with a design first to see how we should support them in a unified way and the task is on me (#38504). Do you mind waiting until the design is done? I'll start to work on it in 2-3 weeks after finishing the Ray 2.7 release. Does this timeline work for you?

File: python/ray/_private/accelerator.py
Comment (allenwang28) [reply python/ray/_private/accelerator.py:187]: Unfortunately not at this moment, this is all mostly internal knowledge related to GKE. I'm not sure if this will be publicized or not

File: python/ray/_private/accelerator.py
Comment (allenwang28) [reply python/ray/_private/accelerator.py:156]: All internal knowledge unfortunately :(

File: python/ray/_private/utils.py
Comment (allenwang28) [reply python/ray/_private/utils.py:479]: I don't disagree, but I will note I was copying the previous examples that Neuron cores and GPU both used.

File: python/ray/_private/accelerator.py
Comment (allenwang28) [reply python/ray/_private/accelerator.py]: Added `FileNotFoundError`!

File: python/ray/_private/accelerator.py
Comment (allenwang28) [reply python/ray/_private/accelerator.py:198]: Done. I moved some logic we used from `config.py` within GCP autoscaler to assert accelerator type (and tested that both are successful)

File: python/ray/_private/ray_constants.py
Comment (allenwang28) [reply python/ray/_private/ray_constants.py:498]: Done

File: python/ray/_private/accelerator.py
Comment (allenwang28) [reply python/ray/_private/accelerator.py]: Done

File: python/ray/_private/ray_option_utils.py
Comment (allenwang28) [reply python/ray/_private/ray_option_utils.py:197]: Done

File: src/ray/common/ray_config_def.h
Comment (allenwang28) [reply src/ray/common/ray_config_def.h]: Great point, I've rewritten the comment to match my understanding

File: python/ray/_private/utils.py
Comment (architkulkarni) [reply python/ray/_private/utils.py:479]: Gotcha! I'm curious why it's there but no need to change it in this PR

**Comment by architkulkarni** [comment]
Thanks, the update looks great!

> Hey @jjyao - I understand, thanks for the heads up. @richardliaw is actually announcing this feature today so I am a bit worried about that timeline of post 2.7. I'm wondering if we can push this version through in time for the 2.7 release? But I can definitely help out with the refactor after the design if you need an extra set of hands!

I'm not familiar with the announcement unfortunately, @richardliaw can you confirm whether or not this PR should be included in Ray 2.7?

**Review by ericl — APPROVED** [review]
Approving for API (not a new API any more).

**Comment by richardliaw** [comment]
hey @architkulkarni looks like approvals are done; can you take a look at tests and merge if tests-ok?

**Comment by architkulkarni** [comment]
Failed tests:
- :kubernetes: ❤️‍🩹 🏃 chaos network delay many job submissions unrelated
- doc build runtime_context.py errors unrelated
- test_client_builder   unrelated
- py38 build failure unrelated

**Comment by matthewdeng** [comment]
This broke the documentation build - @allenwang28 @architkulkarni could you take a look?

```
/home/docs/checkouts/readthedocs.org/user_builds/anyscale-ray/checkouts/38669/python/ray/runtime_context.py:docstring of ray.runtime_context.RuntimeContext.get_resource_ids:5: WARNING: Unexpected indentation.
/home/docs/checkouts/readthedocs.org/user_builds/anyscale-ray/checkouts/38669/python/ray/runtime_context.py:docstring of ray.runtime_context.RuntimeContext.get_resource_ids:3: WARNING: Inline interpreted text or phrase reference start-string without end-string.
```

**Comment by architkulkarni** [comment]
I'll revert this PR and open a new one with the fix.

**Pull request merged**
